### PR TITLE
feat: block registration/sign-in from .shop/.top TLDs

### PR DESCRIPTION
**Pull request by jrf0110**
## Summary

- Adds TLD-based email blocking via a new `BLACKLIST_TLDS` env var (pipe-delimited, e.g. `shop|top`), matching the existing `BLACKLIST_DOMAINS` pattern
- Enforced at the NextAuth `signIn` callback, magic-link API route, and `validateUserAuthorization` for existing sessions
- Includes 8 unit tests covering blocked/allowed TLDs, subdomains, case insensitivity, and edge cases

## Deployment

Set the `BLACKLIST_TLDS` environment variable in production:
```
BLACKLIST_TLDS="foo|bar"
```

### CODE REVIEW

File: src/app/api/auth/magic-link/route.ts [hunk 1]
```diff
@@ -36,6 +37,10 @@ export async function POST(request: NextRequest) {
     return turnstileResult.response;
   }
 
+  if (isEmailBlacklistedByDomain(email) || isBlockedTLD(email)) {
+    return NextResponse.json({ success: false, error: 'BLOCKED' }, { status: 403 });
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/app/api/auth/magic-link/route.ts:41]: Nah

File: src/lib/user.server.ts [hunk 2]
```diff
@@ -764,7 +773,7 @@ async function validateUserAuthorization(
 ): Promise<GetAuthResponse> {
   if (!user) {
     return authError(401, 'User not found', kiloUserId);
-  } else if (isUserBlacklistedByDomain(user)) {
+  } else if (isUserBlacklistedByDomain(user) || isBlockedTLD(user.google_user_email)) {
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/lib/user.server.ts]: Nah

File: src/app/api/auth/magic-link/route.ts [hunk 3]
```diff
@@ -36,11 +37,18 @@ export async function POST(request: NextRequest) {
...
+  }
+
   // Check if this is an existing user (sign-in) or new user (signup)
   const existingUser = await findUserByEmail(email);
 
-  // For new users, enforce stricter email validation
+  // For new users, enforce stricter email validation and TLD blocking
   if (!existingUser) {
+    if (isBlockedTLD(email)) {
+      return NextResponse.json({ success: false, error: 'BLOCKED' }, { status: 403 });
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/app/api/auth/magic-link/route.ts:50]: Nah, I don't think we want to do that

File: src/app/api/auth/magic-link/route.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/app/api/auth/magic-link/route.ts:41]: Already dismissed by repo owner. Resolving.

File: src/lib/user.server.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/lib/user.server.ts]: Already dismissed by repo owner. This code was also removed in the follow-up commit (TLD check no longer runs in validateUserAuthorization). Resolving.

File: src/app/api/auth/magic-link/route.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/app/api/auth/magic-link/route.ts:50]: Already dismissed by repo owner. Resolving.

File: src/lib/user.server.ts [hunk 4]
```diff
@@ -807,6 +823,13 @@ export const isEmailBlacklistedByDomain = (
       email.toLowerCase().endsWith('.' + domain.toLowerCase())
   );
 
+export const isBlockedTLD = (email: string, blacklisted_tlds = BLACKLIST_TLDS) => {
+  const domain = getLowerDomainFromEmail(email);
+  if (!domain) return false;
+  const tld = domain.split('.').pop();
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/lib/user.server.ts]: This feature specifically targets single-part gTLDs like .shop/.top. Multi-part ccTLDs like .co.uk would be handled via BLACKLIST_DOMAINS (full domain matching) if needed — that's the right tool for that job. A comment about this limitation would overstate the scope. Resolving.

File: src/lib/user.server.ts [hunk 5]
```diff
@@ -436,6 +442,16 @@ const authOptions: NextAuthOptions = {
...
+        // Block new signups from blocked TLDs (existing users can still sign in)
+        if (!existingUser && isBlockedTLD(accountInfo.google_user_email)) {
+          sentryLogger('auth', 'warning')(
+            `SECURITY: Blocked TLD signup: ${accountInfo.google_user_email}`,
+            accountInfo
+          );
+
+          return redirectUrlForCode(`BLOCKED`, accountInfo.google_user_email);
+        }
         if (existingUser) {
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/lib/user.server.ts:447]: Fixed in d4c1b5b7 — added blank line for visual separation.

File: src/lib/user.server.ts [hunk 6]
```diff
@@ -84,6 +85,12 @@ const BLACKLIST_DOMAINS = blacklistDomainsEnv
   ? blacklistDomainsEnv.split('|').map((domain: string) => domain.trim())
   : [];
 
+// Pipe-delimited list of TLDs to block, each with a leading dot (e.g. ".shop|.top|.co.uk")
+const blacklistTldsEnv = getEnvVariable('BLACKLIST_TLDS');
+const BLACKLIST_TLDS = blacklistTldsEnv
+  ? blacklistTldsEnv.split('|').map((tld: string) => tld.trim().toLowerCase())
+  : [];
```
Comment (jrf0110) [inline src/lib/user.server.ts]: This needs to be moved to src/lib/config.server.ts

Comment (jrf0110) [reply src/lib/user.server.ts]: Moved to config.server.ts in 576d00a0.

File: src/lib/user.server.ts [hunk 7]
```diff
@@ -436,6 +438,17 @@ const authOptions: NextAuthOptions = {
 
         // Check if this is an existing user with a different primary email
         const existingUser = await findAndSyncExistingUser(accountInfo);
+
+        // Block new signups from blocked TLDs (existing users can still sign in)
+        if (!existingUser && isBlockedTLD(accountInfo.google_user_email)) {
+          sentryLogger('auth', 'warning')(
```
Comment (jrf0110) [inline src/lib/user.server.ts]: Let's not add this noise to sentry

Comment (jrf0110) [reply src/lib/user.server.ts]: Removed the sentryLogger call in 6b947959.

File: src/lib/config.server.ts [hunk 8]
```diff
@@ -186,3 +186,9 @@ export const O11Y_KILO_GATEWAY_CLIENT_SECRET = getEnvVariable('O11Y_KILO_GATEWAY
 export const SECURITY_CLEANUP_BETTERSTACK_HEARTBEAT_URL = getEnvVariable(
   'SECURITY_CLEANUP_BETTERSTACK_HEARTBEAT_URL'
 );
+
+// Pipe-delimited list of TLDs to block from new signups, each with a leading dot (e.g. ".shop|.top|.co.uk")
+const blacklistTldsEnv = getEnvVariable('BLACKLIST_TLDS');
+export const BLACKLIST_TLDS = blacklistTldsEnv
+  ? blacklistTldsEnv.split('|').map((tld: string) => tld.trim().toLowerCase())
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply src/lib/config.server.ts]: Good catch — added .filter(Boolean) to the split in 6b947959.

File: src/lib/config.server.ts [hunk 9]
```diff
@@ -186,3 +186,12 @@ export const O11Y_KILO_GATEWAY_CLIENT_SECRET = getEnvVariable('O11Y_KILO_GATEWAY
 export const SECURITY_CLEANUP_BETTERSTACK_HEARTBEAT_URL = getEnvVariable(
   'SECURITY_CLEANUP_BETTERSTACK_HEARTBEAT_URL'
 );
+
+// Pipe-delimited list of TLDs to block from new signups, each with a leading dot (e.g. ".shop|.top|.co.uk")
```
Comment (markijbema) [inline src/lib/config.server.ts:190]: we're banning the UK?

**Review by markijbema — APPROVED** [review]

File: src/lib/config.server.ts (hunk 9 above)
Comment (jrf0110) [reply src/lib/config.server.ts:190]: Yeah, someone needs to put a stop to their questionable cuisine

Comment (jrf0110) [reply src/lib/config.server.ts:190]: In all seriousness, we are not blocking .co.uk but it was just an example the agent put in there
